File: agentic-clj/test/agentic/trace_test.clj
(ns agentic.trace-test
  "Phase 7: observability. A recording tracer captures spans; the stream runtime opens one span per
   turn and per timer fire, carrying conversation/path/ok attrs and tool: events; the noop default
   keeps behaviour identical when no tracer is attached. Mirrors jagentic-core's trace goldens."
  (:require [clojure.test :refer [deftest is testing]]
            [agentic.core :as core]
            [agentic.event :as ev]
            [agentic.stream :as stream]
            [agentic.timers :as timers]
            [agentic.trace :as trace]))

(deftest recording-tracer-captures-spans
  (testing "start → attr → event → end records one span with name, attrs, events"
    (let [t (trace/recording-tracer)
          span (trace/start t "demo")]
      (-> span (trace/span-attr "k" "v") (trace/span-event "hello") (trace/span-end))
      (let [recorded (trace/spans t)]
        (is (= 1 (count recorded)))
        (let [s (first recorded)]
          (is (= "demo" (:name s)))
          (is (= "v" (get (:attrs s) "k")))
          (is (= ["hello"] (:events s))))
        (is (= ["demo"] (trace/names t)))))))

(deftest stream-traces-turns
  (testing "a traced stream-runtime opens one 'turn' span per event with conversation/path/ok + tools"
    (let [t (trace/recording-tracer)
          sr (-> (stream/stream-runtime (core/banking-system))
                 (stream/with-tracer t))]
      (stream/run sr (stream/seed-channel [(ev/event "c1" "alice" "what is my balance?")]))
      (is (= ["turn"] (trace/names t)))
      (let [s (first (trace/spans t))]
        (is (= "c1" (get (:attrs s) "conversation")))
        (is (= "payments" (get (:attrs s) "path")))
        (is (= "true" (get (:attrs s) "ok")))
        (is (some #{"tool:get_balance"} (:events s))
            "the balance turn invokes the get_balance tool")))))

(deftest timer-fire-traced
  (testing "firing a due timer opens a 'timer.fire' span carrying the payload's path"
    (let [t (trace/recording-tracer)
          sr (-> (stream/stream-runtime (core/banking-system))
                 (stream/with-tracer t))
          ts (timers/in-memory-timer-service)]
      (timers/schedule ts "follow-up" 1000 (ev/event "c1" "alice" "what is my balance?"))
      (stream/fire-due-timers sr ts 1000)
      (is (= ["timer.fire"] (trace/names t)))
      (is (= "payments" (get (:attrs (first (trace/spans t))) "path"))))))

(deftest noop-default-when-no-tracer
  (testing "no with-tracer → noop tracer, behaviour identical (no spans recorded, turns still run)"
    (let [sr (stream/stream-runtime (core/banking-system))
          results (stream/run sr (stream/seed-channel [(ev/event "c1" "alice" "hello there")]))]
      (is (= 1 (count results)))
      (is (= "general" (:path (first results)))))))

File: ports/jagentic-core/src/test/java/org/jagentic/core/trace/TraceTest.java
package org.jagentic.core.trace;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import java.util.List;

import org.junit.jupiter.api.Test;

import org.jagentic.core.Banking;
import org.jagentic.core.ConversationStore;
import org.jagentic.core.Event;
import org.jagentic.core.KeyedStateStore;
import org.jagentic.core.LocalRuntime;
import org.jagentic.core.stream.SeedChannel;
import org.jagentic.core.stream.StreamRuntime;
import org.jagentic.core.timers.InMemoryTimerService;

/** Phase 7: observability. The RecordingTracer captures a span per turn and per timer fire, with
 * the path taken and tool calls as attributes/events. */
class TraceTest {

  private static LocalRuntime banking() {
    return new LocalRuntime(Banking.buildGraph(), new ConversationStore.InMemory(),
        new KeyedStateStore.InMemory(), Banking.defaultTools(), Banking.retriever());
  }

  @Test
  void recordingTracerCapturesAttributesAndEvents() {
    RecordingTracer t = new RecordingTracer();
    t.start("demo").attr("k", "v").event("hello").end();
    assertEquals(1, t.spans().size());
    RecordingTracer.Recorded s = t.spans().get(0);
    assertEquals("demo", s.name());
    assertEquals("v", s.attrs().get("k"));
    assertEquals(List.of("hello"), s.events());
  }

  @Test
  void streamTracesEachTurnWithPathAndTools() {
    RecordingTracer t = new RecordingTracer();
    new StreamRuntime(banking()).withTracer(t)
        .run(new SeedChannel<>(List.of(new Event("c1", "alice", "what is my balance?"))));

    assertEquals(List.of("turn"), t.names());
    RecordingTracer.Recorded turn = t.spans().get(0);
    assertEquals("c1", turn.attrs().get("conversation"));
    assertEquals("payments", turn.attrs().get("path"));
    assertEquals("true", turn.attrs().get("ok"));
    assertTrue(turn.events().contains("tool:get_balance"), "tool call recorded: " + turn.events());
  }

  @Test
  void timerFiresAreTraced() {
    RecordingTracer t = new RecordingTracer();
    StreamRuntime stream = new StreamRuntime(banking()).withTracer(t);
    InMemoryTimerService timers = new InMemoryTimerService();
    timers.schedule("f", 1000, new Event("c1", "alice", "what is my balance?"));
    stream.fireDueTimers(timers, 1000);
    assertEquals(List.of("timer.fire"), t.names());
    assertEquals("payments", t.spans().get(0).attrs().get("path"));
  }

  @Test
  void noopTracerIsTheDefaultAndHarmless() {
    // no withTracer → Tracer.NOOP; run still works, nothing recorded anywhere observable
    var results = new StreamRuntime(banking())
        .run(new SeedChannel<>(List.of(new Event("c1", "u", "hello there"))));
    assertEquals("general", results.get(0).path);
  }
}
